core/txpool/celo_validation.go

```diff
@@ -0,0 +1,133 @@
+func (cvo *CeloValidationOptions) Accepts(txType uint8) bool {
+	_, ok := cvo.AcceptMap[txType]
+	return ok
+}
+
+// CeloValidateTransaction is a helper method to check whether a transaction is valid
+// according to the consensus rules, but does not check state-dependent validation
+// (balance, nonce, etc).
+//
+// This check is public to allow different transaction pools to check the basic
+// rules without duplicating code and running the risk of missed updates.
+func CeloValidateTransaction(tx *types.Transaction, head *types.Header,
+	signer types.Signer, opts *CeloValidationOptions, st *state.StateDB, fcv FeeCurrencyValidator) error {
+
+	if err := ValidateTransaction(tx, head, signer, opts); err != nil {
+		return err
+	}
+	if FeeCurrencyTx(tx) {
+		if !fcv.IsWhitelisted(st, tx.FeeCurrency()) {
+			return NonWhitelistedFeeCurrencyError
+		}
+	}
+	return nil
+}
+
+// NewAcceptMap creates a new AcceptMap with the types provided as keys.
+func NewAcceptMap(types ...uint8) AcceptMap {
+	m := make(AcceptMap, len(types))
+	for _, t := range types {
+		m[t] = struct{}{}
+	}
+	return m
+}
+
+// FeeCurrencyTxType returns true if and only if the transaction type
+// given can handle custom gas fee currencies.
+func FeeCurrencyTxType(t uint8) bool {
+	return t == types.CeloDynamicFeeTxType
+}
+
+// FeeCurrencyTx returns true if this transaction specifies a custom
+// gas fee currency.
+func FeeCurrencyTx(tx *types.Transaction) bool {
+	return FeeCurrencyTxType(tx.Type()) && tx.FeeCurrency() != nil
+}
+
+// See: txpool.ValidationOptionsWithState
+type CeloValidationOptionsWithState struct {
+	State *state.StateDB // State database to check nonces and balances against
+
+	// FirstNonceGap is an optional callback to retrieve the first nonce gap in
+	// the list of pooled transactions of a specific account. If this method is
+	// set, nonce gaps will be checked and forbidden. If this method is not set,
+	// nonce gaps will be ignored and permitted.
+	FirstNonceGap func(addr common.Address) uint64
+
+	// UsedAndLeftSlots is a mandatory callback to retrieve the number of tx slots
+	// used and the number still permitted for an account. New transactions will
+	// be rejected once the number of remaining slots reaches zero.
+	UsedAndLeftSlots func(addr common.Address) (int, int)
+
+	// ExistingExpenditure is a mandatory callback to retrieve the cummulative
+	// cost of the already pooled transactions to check for overdrafts.
+	ExistingExpenditure func(addr common.Address) *big.Int
+
+	// ExistingCost is a mandatory callback to retrieve an already pooled
+	// transaction's cost with the given nonce to check for overdrafts.
+	ExistingCost func(addr common.Address, nonce uint64) *big.Int
+
+	// L1CostFn is an optional extension, to validate L1 rollup costs of a tx
+	L1CostFn L1CostFunc
+
+	// Celo
+
+	// FeeCurrencyValidator allows for balance check of non native fee currencies.
+	FeeCurrencyValidator FeeCurrencyValidator
+}
```

Comment:
Why repeat the struct instead of extending the existing one?
```suggestion
type CeloValidationOptionsWithState struct {
    txpool.ValidationOptionsWithState
    
	// FeeCurrencyValidator allows for balance check of non native fee currencies.
	FeeCurrencyValidator FeeCurrencyValidator
}
```

Reply:
I thought about that and I was a little bit torn between the two options. 1) The one I wrote, a bit annoying to replicate the fields in the celo file, but allows minimum modification of the upstream file. 2) Your suggestion, which is more concise but (I used to think) provokes several line changes (mostly indentation). But re-thinking it I think I have a way to prevent the several indentation changes. I'll change it.